modules/xmpp/xmpp.js

```diff
@@ -490,50 +496,62 @@ export default class XMPP extends Listenable {
-        // presence at the top of the send queue. We flush() once more after
-        // disconnect() in order to attempt to have its unavailable presence
-        // sent as soon as possible.
-        this.connection.flush();
-
-        if (ev !== null && typeof ev !== 'undefined') {
-            const evType = ev.type;
-
-            if (evType === 'beforeunload' || evType === 'unload') {
-                // XXX Whatever we said above, synchronous sending is the best
-                // (known) way to properly disconnect from the XMPP server.
-                // Consequently, it may be fine to have the source code and
-                // comment it in or out depending on whether we want to run with
-                // it for some time.
-                this.connection.options.sync = true;
+        return new Promise(resolve => {
+            const disconnectListener = (credentials, status) => {
+                if (status === Strophe.Status.DISCONNECTED) {
+                    resolve();
+                    this.eventEmitter.removeListener(XMPPEvents.CONNECTION_STATUS_CHANGED, disconnectListener);
+                }
+            };
+
+            this.eventEmitter.on(XMPPEvents.CONNECTION_STATUS_CHANGED, disconnectListener);
+
+            // XXX Strophe is asynchronously sending by default. Unfortunately, that
+            // means that there may not be enough time to send an unavailable
+            // presence or disconnect at all. Switching Strophe to synchronous
+            // sending is not much of an option because it may lead to a noticeable
+            // delay in navigating away from the current location. As a compromise,
+            // we will try to increase the chances of sending an unavailable
+            // presence and/or disconecting within the short time span that we have
+            // upon unloading by invoking flush() on the connection. We flush() once
+            // before disconnect() in order to attemtp to have its unavailable
+            // presence at the top of the send queue. We flush() once more after
+            // disconnect() in order to attempt to have its unavailable presence
+            // sent as soon as possible.
+            this.connection.flush();
+
+            if (ev !== null && typeof ev !== 'undefined') {
+                const evType = ev.type;
+
+                if (evType === 'beforeunload' || evType === 'unload') {
+                    // XXX Whatever we said above, synchronous sending is the best
+                    // (known) way to properly disconnect from the XMPP server.
+                    // Consequently, it may be fine to have the source code and
+                    // comment it in or out depending on whether we want to run with
+                    // it for some time.
+                    this.connection.options.sync = true;
+                }
             }
-        }
 
-        this.connection.disconnect();
+            this.connection.disconnect();
```

Comment:
To confirm, this disconnect call doesn't return a promise that can be re-used? The way the

Reply:
According to strophe's docs and their current source code there isn't a promise. At the beginning I was also hoping there is :) but unfortunately ...